src/streamer.rs

```diff
@@ -473,77 +495,120 @@ fn recv_window(
             *received,
             consume_queue.len(),
         );
-        trace!("sending consume_queue.len: {}", consume_queue.len());
+        trace!(
+            "{:x}: sending consume_queue.len: {}",
+            debug_id,
+            consume_queue.len()
+        );
         inc_new_counter!("streamer-recv_window-consume", consume_queue.len());
         s.send(consume_queue)?;
     }
     Ok(())
 }
 
-fn print_window(debug_id: u64, locked_window: &Window, consumed: u64) {
-    {
-        let buf: Vec<_> = locked_window
-            .read()
-            .unwrap()
-            .iter()
-            .enumerate()
-            .map(|(i, v)| {
-                if i == (consumed % WINDOW_SIZE) as usize {
-                    "_"
-                } else if v.is_none() {
-                    "0"
-                } else if let Some(ref cblob) = v {
-                    if cblob.read().unwrap().is_coding() {
-                        "C"
-                    } else {
-                        "1"
-                    }
-                } else {
-                    "0"
-                }
-            })
-            .collect();
-        trace!("{:x}:WINDOW ({}): {}", debug_id, consumed, buf.join(""));
-    }
+fn print_window(debug_id: u64, window: &Window, consumed: u64) {
```

Comment:
Neato. How about returning that string and using it for tests?

Reply:
I'm actually a little concerned that this code may execute when not needed (i.e. when RUST_LOG doesn't have solana::streamer=...) constructing a 2kb string every time we get a packet can't be good for performance.  can I check whether... fixed now, I think...

Reply:
https://docs.rs/log/0.4.3/log/macro.log_enabled.html